### ISSUE TITLE
Wrong implementation of Guided Backprop ?

### ISSUE THREAD

**Issue by homaralex** [utkuozbulak/pytorch-cnn-visualizations#36]
Isn't there a bug in the way negative gradients are filtered out for ReLU's in Guided Backprop?

What you do is:
`torch.clamp(grad_in[0], min=0.0)`, 

while the paper states:

> rather than masking out values corresponding to negative entries of the top gradient (’deconvnet’) or bottom data (backpropagation), we mask out the values for which at least one
of these values is negative

It seems to me that you forgot to include the top gradient.

**Comment by utkuozbulak** [comment]
>  We call this method guided backpropagation, because it adds an additional guidance signal from the higher layers to usual backpropagation. **This prevents backward flow of negative gradients,** corresponding to the neurons which decrease the activation of the higher layer unit we aim to visualize.

Am I missing something?

**Comment by homaralex** [comment]
I think you should also be masking the gradients wrt. output (top) gradients, in a similar manner like [here](https://github.com/experiencor/deep-viz-keras/blob/master/guided_backprop.py)

**Comment by utkuozbulak** [comment]
Hey, you mean the gradients should also be filtered by non-zero forward output of the layers? Having re-read the paper (especially in Figure1), I get what you mean. In fact, the implemented approach (in the repository) looks closer to deconvolution (Zeiler&Fergus).

I will go over it during the weekend, there are few more things I want to change so I might as well implement them all together.

Thanks.
